source/tutorials/nix-shell-in-shebang.md

```diff
@@ -0,0 +1,76 @@
+# Reproducible interpreted scripts
+
+In this tutorial, you will learn how to use Nix to create and run reproducible interpreted scripts with [shebang].
+
+## Requirements
+
+- A working [Nix installation](install-nix)
+- Familiarity with [Bash]
+
+## A trivial script with non-trivial dependencies
+
+Take the following script, which fetches the content XML of a URL, converts it to JSON, and formats it for better readability:
+
+```bash
+#! /bin/bash
+
+curl https://github.com/NixOS/nixpkgs/releases.atom | xml2json | jq .
+```
+
+It requires the programs `curl`, `xml2json`, and `jq`.
+It also requires the `bash` interpreter.
+If any of these dependencies are not present on the system running the script, it will fail partially or altogether.
+
+With Nix, we can declare all dependencies explicitly, and produce a script that will always run on any machine that supports Nix and the required packages taken from Nixpkgs.
+
+## The script
+
+A [shebang] is the first line of a script starting with `#!`.
+It determines which program to use for running the script.
+
+[Bash]: https://www.gnu.org/software/bash/
+[shebang]: https://en.wikipedia.org/wiki/Shebang_(Unix)
+
+We will use the shebang line `#! /usr/bin/env nix-shell`.
+
+`/usr/bin/env` is a program available on most modern Unix-like operating systems.
+It takes a command name as argument and will run the first executable by that name it finds in the directories listed in the environment variable `$PATH`.
+
+We use [`nix-shell` as a shebang interpreter].
+It takes the following parameters relevant for our use case:
+
+[`nix-shell` as a shebang interpreter]: https://nixos.org/manual/nix/stable/command-ref/nix-shell.html#use-as-a--interpreter
+- `-i` tells which program to use for interpreting the rest of the file
+- `-p` lists packages that should be present in the interpreter's environment
+- `-I` explicitly sets the search path for packages
+- `--pure` clears the environment before start - more details in the [nix-shell man page]: https://nixos.org/manual/nix/stable/command-ref/nix-shell.html#options
```

Comment:
Purely for stylistic reasons, I would remove `--pure` here and make a separate section for it. I think it is the single most useful insurance for portability in `nix-shell` (and shockingly missing from `nix shell`, whose `-i` flag doesn't clear `PATH`!). If you feel like this is too much effort or undesirable I'm happy to make a separate PR with this modification once this is merged.

I was basically envisioning this tutorial and script minus `--pure`, and then discussion and an example of how esp on non-NixOS systems your `PATH` is still being search for these binaries (that might not be present on someone else's system). Purity is the sine qua non of Nix IMO and I think deserves special attention for this *really important* use case!

Thanks for considering.

Reply:
I can see the value in the suggestion Nathan. For me it's a difficult section to write as my understanding isn't that advanced yet, so I'd rather pass on including it for now and make a separate revision/pull request for it. 

I have a similar thing that I don't understand but could also be good to include: With the I-flag, how can I figure out what the eligible values are to put in for the search path? I am not in a position to write that either as I don't know the answer.

Reply:
Sounds good, I'll make a separate PR with an example.

Sorry, you're referring to `-I explicitly sets the search path for packages`?

That wording should maybe be changed? I'm not that advanced either, but my understanding is that sets the search path for looking up e.g. `nixpkgs`, not for "packages" in the traditional sense.

For example, the first two of these give me the same binary of nixpkgs because using my user's default search path for `nixpkgs` is implicit in the first and explicit in the second. In the third, I override `nixpkgs` with a path to a locally checked out repository using a different version of nixpkgs (current master, recently pulled). In the fourth invocation, I pinned a specific version of `nixpkgs`, which should help ensure complete reproducibility.

So the packages you get depend entirely on the version of `nixpkgs` you're setting in your channel.

(Flakes are relevant here, as they are always explicit about the version of `nixpkgs` being used.)

```console
$ nix-shell --pure -p nix --command 'type -a nix; nix --version'
nix is /nix/store/gzd5wpvq9b5q5l3hvwpihgrhcrrllgi0-nix-2.11.1/bin/nix
nix (Nix) 2.11.1
$ nix-shell --pure -p nix -I nixpkgs=~/.nix-defexpr/channels/nixpkgs --command 'type -a nix; nix --version'
nix is /nix/store/gzd5wpvq9b5q5l3hvwpihgrhcrrllgi0-nix-2.11.1/bin/nix
nix (Nix) 2.11.1
$ nix-shell --pure -p nix -I nixpkgs=~/git/nixpkgs --command 'type -a nix; nix --version'
nix is /nix/store/gmknmx6vrdj8bzzfmyk9qfcalhsrpc3a-nix-2.12.0/bin/nix
nix (Nix) 2.12.0
$ nix-shell --pure -p nix -I nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/92a047a6c4d46a222e9c323ea85882d0a7a13af8.tar.gz --command 'type -a nix; nix --version' 
nix is /nix/store/w1xw3nsq03m2h8c0g23ps62lcbkdlc83-nix-2.0.4/bin/nix
nix (Nix) 2.0.4
```

For the record, if you want a list of all packages in that revision of `nixpkgs`, something like is sloppy but seems to work (I'm sure others can tell you a better way.

```console
$ nix-instantiate --eval -E 'with (import <nixpkgs> {}); lib.attrNames (import "${pkgs.path}/pkgs/top-level/all-packages.nix" { \inherit lib config overlays; noSysDirs = true; } {} {} {})'
$ # alternatively, with new-style syntax and json output
$ nix eval --impure --json --expr 'with (import <nixpkgs> {}); lib.attrNames (import "${pkgs.path}/pkgs/top-level/all-packages.nix" { inherit lib config overlays; noSysDirs = true; } {} {} {})' | jq
```

Or just `nix eval nixpkgs#<TAB><TAB>`.

Reply:
Thank you for sharing the examples above - this is really instructive. 

Two follow-up questions. My standard shell is zsh instead of bash.

For some reason, for example two in zsh I get back:
`warning: Nix search path entry '~/.nix-defexpr/channels/nixpkgs/' does not exist, ignoring`
whereas if I enter bash, and run the very same command, it works. 
It might not be a nix-shell question strictly, but I would love to know why it behaves differently in zsh and bash.

In the fourth example, how did you work out that that specific archive tar.gz file would exist? Can you browse these on github, or is there some other way? 

Thanks a lot!